File: flutter_app/lib/models/inventory_models.dart
class InventoryItem {
  final String id;
  final String name;
  final String category;
  final String quantity;
  final double fullness; // 0.0 to 1.0
  final String? imagePath; // local path to stylized image
  final String? brand;
  final String? notes;
  final DateTime addedDate;
  final DateTime lastUpdated;
  final bool expiresSoon;

  InventoryItem({
    required this.id,
    required this.name,
    required this.category,
    required this.quantity,
    required this.fullness,
    this.imagePath,
    this.brand,
    this.notes,
    required this.addedDate,
    required this.lastUpdated,
    this.expiresSoon = false,
  });

  factory InventoryItem.fromJson(Map<String, dynamic> json) {
    return InventoryItem(
      id: json['id'],
      name: json['name'],
      category: json['category'],
      quantity: json['quantity'],
      fullness: (json['fullness'] ?? 1.0).toDouble(),
      imagePath: json['image_path'],
      brand: json['brand'],
      notes: json['notes'],
      addedDate: DateTime.parse(json['added_date']),
      lastUpdated: DateTime.parse(json['last_updated']),
      expiresSoon: json['expires_soon'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'category': category,
      'quantity': quantity,
      'fullness': fullness,
      'image_path': imagePath,
      'brand': brand,
      'notes': notes,
      'added_date': addedDate.toIso8601String(),
      'last_updated': lastUpdated.toIso8601String(),
      'expires_soon': expiresSoon,
    };
  }

  InventoryItem copyWith({
    String? id,
    String? name,
    String? category,
    String? quantity,
    double? fullness,
    String? imagePath,
    String? brand,
    String? notes,
    DateTime? addedDate,
    DateTime? lastUpdated,
    bool? expiresSoon,
  }) {
    return InventoryItem(
      id: id ?? this.id,
      name: name ?? this.name,
      category: category ?? this.category,
      quantity: quantity ?? this.quantity,
      fullness: fullness ?? this.fullness,
      imagePath: imagePath ?? this.imagePath,
      brand: brand ?? this.brand,
      notes: notes ?? this.notes,
      addedDate: addedDate ?? this.addedDate,
      lastUpdated: lastUpdated ?? this.lastUpdated,
      expiresSoon: expiresSoon ?? this.expiresSoon,
    );
  }
}

class RecognizedIngredient {
  final String name;
  final String category;
  final double confidence;
  final String? brand;
  final String? quantityEstimate;
  final String? locationDescription;

  RecognizedIngredient({
    required this.name,
    required this.category,
    required this.confidence,
    this.brand,
    this.quantityEstimate,
    this.locationDescription,
  });

  factory RecognizedIngredient.fromJson(Map<String, dynamic> json) {
    return RecognizedIngredient(
      name: json['name'],
      category: json['category'],
      confidence: json['confidence'].toDouble(),
      brand: json['brand'],
      quantityEstimate: json['quantity_estimate'],
      locationDescription: json['location_description'],
    );
  }
}

class ImageRecognitionResponse {
  final List<RecognizedIngredient> recognizedIngredients;
  final List<String> suggestions;
  final double? processingTime;

  ImageRecognitionResponse({
    required this.recognizedIngredients,
    required this.suggestions,
    this.processingTime,
  });

  factory ImageRecognitionResponse.fromJson(Map<String, dynamic> json) {
    var ingredientsList = json['recognized_ingredients'] as List;
    List<RecognizedIngredient> ingredients = ingredientsList
        .map((item) => RecognizedIngredient.fromJson(item))
        .toList();

    return ImageRecognitionResponse(
      recognizedIngredients: ingredients,
      suggestions: List<String>.from(json['suggestions']),
      processingTime: json['processing_time']?.toDouble(),
    );
  }
}

class InventoryStats {
  final int totalItems;
  final Map<String, int> byCategory;
  final Map<String, int> byQuantity;
  final int expiringSoon;
  final DateTime lastUpdated;

  InventoryStats({
    required this.totalItems,
    required this.byCategory,
    required this.byQuantity,
    required this.expiringSoon,
    required this.lastUpdated,
  });

  factory InventoryStats.fromJson(Map<String, dynamic> json) {
    return InventoryStats(
      totalItems: json['total_items'],
      byCategory: Map<String, int>.from(json['by_category']),
      byQuantity: Map<String, int>.from(json['by_quantity']),
      expiringSoon: json['expiring_soon'],
      lastUpdated: DateTime.parse(json['last_updated']),
    );
  }
}

class IngredientCategory {
  static const String spirits = 'spirits';
  static const String liqueurs = 'liqueurs';
  static const String bitters = 'bitters';
  static const String syrups = 'syrups';
  static const String juices = 'juices';
  static const String freshIngredients = 'fresh_ingredients';
  static const String garnishes = 'garnishes';
  static const String mixers = 'mixers';
  static const String equipment = 'equipment';
  static const String other = 'other';

  static List<String> get all => [
        spirits,
        liqueurs,
        bitters,
        syrups,
        juices,
        freshIngredients,
        garnishes,
        mixers,
        equipment,
        other,
      ];

  static String getDisplayName(String category) {
    switch (category) {
      case spirits:
        return 'Spirits';
      case liqueurs:
        return 'Liqueurs';
      case bitters:
        return 'Bitters';
      case syrups:
        return 'Syrups';
      case juices:
        return 'Juices';
      case freshIngredients:
        return 'Fresh Ingredients';
      case garnishes:
        return 'Garnishes';
      case mixers:
        return 'Mixers';
      case equipment:
        return 'Equipment';
      case other:
        return 'Other';
      default:
        return category;
    }
  }
}

class QuantityDescription {
  static const String empty = 'empty';
  static const String almostEmpty = 'almost_empty';
  static const String quarterBottle = 'quarter_bottle';
  static const String halfBottle = 'half_bottle';
  static const String threeQuarterBottle = 'three_quarter_bottle';
  static const String fullBottle = 'full_bottle';
  static const String multipleBottles = 'multiple_bottles';
  static const String smallAmount = 'small_amount';
  static const String mediumAmount = 'medium_amount';
  static const String largeAmount = 'large_amount';
  static const String veryLargeAmount = 'very_large_amount';

  static List<String> get all => [
        empty,
        almostEmpty,
        quarterBottle,
        halfBottle,
        threeQuarterBottle,
        fullBottle,
        multipleBottles,
        smallAmount,
        mediumAmount,
        largeAmount,
        veryLargeAmount,
      ];

  static String getDisplayName(String quantity) {
    switch (quantity) {
      case empty:
        return 'Empty';
      case almostEmpty:
        return 'Almost Empty';
      case quarterBottle:
        return 'Quarter Bottle';
      case halfBottle:
        return 'Half Bottle';
      case threeQuarterBottle:
        return 'Three Quarter Bottle';
      case fullBottle:
        return 'Full Bottle';
      case multipleBottles:
        return 'Multiple Bottles';
      case smallAmount:
        return 'Small Amount';
      case mediumAmount:
        return 'Medium Amount';
      case largeAmount:
        return 'Large Amount';
      case veryLargeAmount:
        return 'Very Large Amount';
      default:
        return quantity;
    }
  }

  static double getFullnessValue(String quantity) {
    switch (quantity) {
      case empty:
        return 0.0;
      case almostEmpty:
        return 0.1;
      case quarterBottle:
        return 0.25;
      case halfBottle:
        return 0.5;
      case threeQuarterBottle:
        return 0.75;
      case fullBottle:
        return 1.0;
      case multipleBottles:
        return 1.0; // Could be > 1.0 but we'll cap at 1.0 for display
      case smallAmount:
        return 0.2;
      case mediumAmount:
        return 0.5;
      case largeAmount:
        return 0.8;
      case veryLargeAmount:
        return 1.0;
      default:
        return 1.0;
    }
  }
}

File: mixologist/models/inventory_models.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Union
from enum import Enum
from datetime import datetime


class QuantityDescription(str, Enum):
    """User-friendly quantity descriptions."""
    EMPTY = "empty"
    ALMOST_EMPTY = "almost empty"
    QUARTER_BOTTLE = "quarter bottle"
    HALF_BOTTLE = "half bottle"
    THREE_QUARTER_BOTTLE = "three quarter bottle"
    FULL_BOTTLE = "full bottle"
    MULTIPLE_BOTTLES = "multiple bottles"
    SMALL_AMOUNT = "small amount"
    MEDIUM_AMOUNT = "medium amount"
    LARGE_AMOUNT = "large amount"
    VERY_LARGE_AMOUNT = "very large amount"


class IngredientCategory(str, Enum):
    """Categories of cocktail ingredients."""
    SPIRITS = "spirits"
    LIQUEURS = "liqueurs" 
    BITTERS = "bitters"
    SYRUPS = "syrups"
    JUICES = "juices"
    FRESH_INGREDIENTS = "fresh_ingredients"
    GARNISHES = "garnishes"
    MIXERS = "mixers"
    EQUIPMENT = "equipment"
    OTHER = "other"


class InventoryItem(BaseModel):
    """Individual inventory item."""
    id: str = Field(..., description="Unique identifier for the item")
    name: str = Field(..., description="Name of the ingredient")
    category: IngredientCategory = Field(..., description="Category of the ingredient")
    quantity: QuantityDescription = Field(..., description="User-friendly quantity description")
    brand: Optional[str] = Field(None, description="Brand name if applicable")
    notes: Optional[str] = Field(None, description="Additional user notes")
    added_date: datetime = Field(default_factory=datetime.now, description="When item was added")
    last_updated: datetime = Field(default_factory=datetime.now, description="Last update timestamp")
    expires_soon: Optional[bool] = Field(False, description="Flag if ingredient expires soon")
    
    def update_quantity(self, new_quantity: QuantityDescription):
        """Update quantity and timestamp."""
        self.quantity = new_quantity
        self.last_updated = datetime.now()


class InventoryAddRequest(BaseModel):
    """Request to add an item to inventory."""
    name: str = Field(..., description="Name of the ingredient")
    category: IngredientCategory = Field(..., description="Category of the ingredient")
    quantity: QuantityDescription = Field(..., description="Quantity description")
    brand: Optional[str] = Field(None, description="Brand name if applicable")
    notes: Optional[str] = Field(None, description="Additional notes")


class InventoryUpdateRequest(BaseModel):
    """Request to update an inventory item."""
    quantity: Optional[QuantityDescription] = Field(None, description="New quantity")
    brand: Optional[str] = Field(None, description="New brand")
    notes: Optional[str] = Field(None, description="New notes")
    expires_soon: Optional[bool] = Field(None, description="Expiry flag")


class ImageRecognitionRequest(BaseModel):
    """Request for OpenAI vision analysis of inventory image."""
    image_base64: str = Field(..., description="Base64 encoded image")
    existing_inventory: Optional[List[str]] = Field([], description="List of existing inventory items")


class RecognizedIngredient(BaseModel):
    """Ingredient recognized from image."""
    name: str = Field(..., description="Recognized ingredient name")
    category: IngredientCategory = Field(..., description="Predicted category")
    confidence: float = Field(..., description="Confidence score (0-1)")
    brand: Optional[str] = Field(None, description="Recognized brand if visible")
    quantity_estimate: Optional[QuantityDescription] = Field(None, description="Estimated quantity if visible")
    location_description: Optional[str] = Field(None, description="Where in image this item was found")


class ImageRecognitionResponse(BaseModel):
    """Response from image recognition."""
    recognized_ingredients: List[RecognizedIngredient] = Field(..., description="List of recognized ingredients")
    suggestions: List[str] = Field([], description="Additional suggestions or notes")
    processing_time: Optional[float] = Field(None, description="Time taken to process")


class InventoryFilterRequest(BaseModel):
    """Request to filter recipes based on inventory."""
    available_only: bool = Field(True, description="Only show recipes with available ingredients")
    include_substitutions: bool = Field(True, description="Include recipes with ingredient substitutions")
    missing_ingredient_limit: int = Field(0, description="Max missing ingredients to allow")


class InventoryStats(BaseModel):
    """Statistics about current inventory."""
    total_items: int = Field(..., description="Total number of items")
    by_category: Dict[str, int] = Field(..., description="Count by category")
    by_quantity: Dict[str, int] = Field(..., description="Count by quantity level")
    expiring_soon: int = Field(..., description="Items expiring soon")
    last_updated: datetime = Field(..., description="When inventory was last updated")


class Inventory(BaseModel):
    """Complete inventory structure."""
    items: List[InventoryItem] = Field(default_factory=list, description="List of inventory items")
    user_id: str = Field("default_user", description="User identifier")
    created_date: datetime = Field(default_factory=datetime.now, description="When inventory was created")
    last_updated: datetime = Field(default_factory=datetime.now, description="Last update timestamp")
    
    def add_item(self, item: InventoryItem) -> None:
        """Add an item to inventory."""
        self.items.append(item)
        self.last_updated = datetime.now()
    
    def get_item_by_id(self, item_id: str) -> Optional[InventoryItem]:
        """Get item by ID."""
        return next((item for item in self.items if item.id == item_id), None)
    
    def remove_item(self, item_id: str) -> bool:
        """Remove item by ID."""
        original_length = len(self.items)
        self.items = [item for item in self.items if item.id != item_id]
        if len(self.items) < original_length:
            self.last_updated = datetime.now()
            return True
        return False
    
    def get_stats(self) -> InventoryStats:
        """Get inventory statistics."""
        by_category = {}
        by_quantity = {}
        expiring_soon = 0
        
        for item in self.items:
            # Count by category
            category_key = item.category.value
            by_category[category_key] = by_category.get(category_key, 0) + 1
            
            # Count by quantity
            quantity_key = item.quantity.value
            by_quantity[quantity_key] = by_quantity.get(quantity_key, 0) + 1
            
            # Count expiring items
            if item.expires_soon:
                expiring_soon += 1
        
        return InventoryStats(
            total_items=len(self.items),
            by_category=by_category,
            by_quantity=by_quantity,
            expiring_soon=expiring_soon,
            last_updated=self.last_updated
        )